Encapsulation and ECIES (v2)

athoelke opened the pull request:
Starting from #220 - this PR:

* Strips out ML-KEM (to be presented in a PQC extension document).
* Updates the text align with other key establishment APIs.
* Provide a more general overview for KEMs.
* Provide a detailed alignment for ECIES with the SEC1 procedures, as the ECIES in the Crypto API is just the key agreement aspect.
* Moved some of the description and commentary into admonitions to clarify the content.

~~Some open issues:~~

* ~~The encapsulation data sizing macro names. On other APIs, we typically name these after the function and the output parameter - but that would result in `PSA_ENCAPSULATE_ENCAPSULATION_SIZE()` which seems somewhat repetitive. I have currently gone with `PSA_ENCAPSULATION_SIZE()`, as the variable name is pretty distinctive.~~
* ~~Should we consistently call this 'Encapsulation', 'Key encapsulation', or 'Key encapsulation mechanism'? - currently the text is mostly a mix of the first two.~~
* ~~The output key is variously referred to as 'output key', 'shared output key', 'shared secret', and 'shared secret key'. Is there a best/preferred term for this in this API?~~
* ~~If a decapsulation algorithm has an explicit failure (returns an error if the decapsulation fails rather than the wrong key), what error code should we use? (INVALID_ARGUMENT, or INVALID_SIGNATURE, or something else?).~~

This PR provides the API foundations for ML-KEM (#95).

athoelke commented:
Another thing I am noticing in the additions for 1.3 - there are a lot of new key-returning functions. The list of 'key creation' functions is repeated about 8 times in the spec, and these are getting quite chunky now with over 10 such functions.

It might be worth reviewing those passages and simplifying the wording, and references to ALL of these APIs.

athoelke commented:
Updated:

* Decided to use "key-encapsulation algorithm" everywhere
* Use "shared secret" when describing the result value of the encapsulation/decapsulation process
* Use "shared output key" when describing the key object returned by `psa_encapsulate()` and `psa_decapsulate()`
* Permit `PSA_ERROR_INVALID_SIGNATURE` for encapsulation data authentication failure
* Describe failure possibility for probablistic algorithms (e.g. ML-KEM) as reason for non-identical output keys
* Warn implementations against reporting padding validity errors

athoelke commented:
Decided that `PSA_ENCAPSULATION_SIZE()` and `PSA_ENCAPSULATION_MAX_SIZE` are suitable for the support macro names.

gilles-peskine-arm commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+    *   The shared secret :math:`Z` is output as the shared output key.
+
+    The encapsulation data produced by `PSA_ALG_ECIES_SEC1` is not authenticated.
+    When used in a full ECIES scheme, the authentication of the encrypted message implicitly confirms that the derived keys were identical.
```
> the authentication of the encrypted message implicitly confirms that the derived keys were identical

Does it? Isn't this assuming that the authentication is key-committing, which is generally the case for hash-based MACs but not for other common authentication algorithms?

athoelke replied:
So perhaps the wording is too specific about what the AE scheme, or cipher+MAC scheme ensures.

The gist of what I wanted to say was that explicit confirmation of the shared secret key is not used in ECIES. The scheme uses either a keyed MAC or authenticated encryption using a key derived from the shared secret to authenticate the encrypted message.

athoelke replied:
I've reworded this to something more specific to ECIES and its purpose:

> In the full ECIES scheme, the authentication of the encrypted message using a key derived from the shared secret provides assurance that the message has not been manipulated.

gilles-peskine-arm reviewed (changes requested):
Looks mostly good to me, but with a few reservations, mostly about terminology.

athoelke commented on doc/crypto/api.db/psa/crypto.h:
```diff
@@ -498,7 +505,20 @@ psa_status_t psa_copy_key(psa_key_id_t source_key,
 psa_status_t psa_crypto_init(void);
+psa_status_t psa_decapsulate(psa_key_id_t key,
+                             psa_algorithm_t alg,
+                             const uint8_t * encapsulation,
```
I think I was trying to use a term that was 'kind of' appropriate for the data exchanged for the key establishment phase of ECIES. Technically it is not ciphertext, but given this API is more specifically geared toward KEMs, I am happy to switch to 'ciphertext' as the common term here.

athoelke replied:
... and deal with the oddness of calling the public key share that ECIES outputs for the key establishment the 'ciphertext'.

athoelke commented on doc/crypto/api/keys/policy.rst:
```diff
@@ -250,6 +261,31 @@ The usage flags are encoded in a bitmask, which has the type `psa_key_usage_t`.
+    .. summary::
+        Permission to encapsulate new keys.
+
+    This flag is required to encapsulate new keys to send to a counter party.
```
Key encapsulation is always about establishing a 'new' key. That said, this text is not consistent with the other usage flags. _If_ we keep this (see comment on L99), then it might be better worded as:
 > Permission to perform key encapsulation with the key.

athoelke commented on doc/crypto/api/keys/policy.rst:
```diff
@@ -250,6 +261,31 @@ The usage flags are encoded in a bitmask, which has the type `psa_key_usage_t`.
+    .. summary::
+        Permission to encapsulate new keys.
+
+    This flag is required to encapsulate new keys to send to a counter party.
```
I missed these uses in polict.rst when replacing '[counter] party' with 'participant'

athoelke commented on doc/crypto/api/keys/types.rst:
```diff
@@ -806,10 +806,16 @@ The curve type affects the key format, the key derivation procedure, and the alg
+
             *   -   Montgomery
-                -   Montgomery curve key-pairs can only be used in key agreement algorithms.
+                -   Montgomery curve key-pairs can be used in key agreement and key-encapsulation algorithms.
```
Some data points:

* FIPS205 consistently uses "key-encapsulation mechanisms".
* In other documents, NIST uses "key-establishment scheme", "key-transport scheme", "key-agreement scheme" - where key-xxx is a noun-modifier - and no hypen when "key establishment", "key agreement" etc is the noun.
* Wikipedia has "key encapsulation mechanism", but also "key-agreement protocol".

I think the form used by NIST is the more generally accepted approach to use of hyphens in compound noun expressions. So perhaps I should revise the existing text where we use "key agreement algorithm" or "key establishment procedure", etc.?

athoelke replied:
I'm now planning to adjust the hyphenation of key-agreement terms across the specification (in a different PR).

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+The encapsulation data is transferred to the first participant, who uses the private decapsulation key to compute another copy of the shared secret.
+
+Typically, the shared secret is used as input to a key-derivation function, to create keys for secure communication between the participants.
+However, some key-encapsulation algorithms result in a pseudo-random shared secret, which is suitable to be used directly as a cryptographic key.
```
Thank you.

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+.. function:: psa_encapsulate
+
+    .. summary::
+        Use a public key to generate a new shared secret key and associated encapsulation data.
```
Happy to go with 'ciphertext'.

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+
+        *   The key type.
+            All key-encapsulation algorithms can output a key of type :code:`PSA_KEY_TYPE_DERIVE` or :code:`PSA_KEY_TYPE_HMAC`.
+            Key-encapsulation algorithms that produce a pseudo-random shared secret, can also output block-cipher key types, for example :code:`PSA_KEY_TYPE_AES`.
```
This wording was duplicated from the PAKE `psa_pake_get_shared_key()` API, which has similar properties, depending on the PAKE algorithm in use. Some PAKEs ensure that the shared secret is uniformly pseudorandom, and of a useful size for keying material. So we worded the API description to permit immediate use as a key to permit use cases where the higher level protocol does not do any further key derivation.

I don't think we can rule out such higher level protocol designs for encapsulation, given the following text from FIPS203 (ML-KEM):

> **Approved usage of the shared secret key.** If randomness generation is successful, the values *K* and *K'* returned by ML-KEM.Encaps and ML-KEM.Decaps, respectively, are always 256-bit values. Under appropriate conditions (see Sections 3.1 and 3.2, and SP 800-227), these values match (i.e., *K = K'*) and can be used directly as a shared secret key for symmetric cryptography. If further key derivation is needed, the final symmetric keys shall be derived from this 256-bit shared secret key in an approved manner, as specified in SP 800-108 and SP 800-56C.

athoelke replied:
There is already a statement to direct the reader to the individual key-encapsulation algorithm documentation for specific details.

gilles-peskine-arm replied:
AFAIK it's not that common to directly use a PAKE output as a “proper” key, rather than a verification value or fodder for derivation. On the other hand, it's pretty common after a key agreement to want to do minimal work to use the shared secret as an AEAD key. That's only secure if the shared secret is uniformly pseudorandom (which is not the case for the dominant key agreement, namely FFDH/ECDH), and that tends to require truncation. For a KEM designed to have uniformly pseudorandom output, I expect protocols will specify things like “use the first 128 bits as the AES key”.

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+
+        The following attributes are optional:
+
+        *   If the key size is nonzero, it must be equal to the size of the shared secret.
```
Good point. Will clarify that this is the bit-size of the shared secret.

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+    .. retval:: PSA_ERROR_BUFFER_TOO_SMALL
+        The size of the ``encapsulation`` buffer is too small.
+        `PSA_ENCAPSULATION_SIZE()` or `PSA_ENCAPSULATION_MAX_SIZE` can be used to determine a sufficient buffer size.
+    .. retval:: PSA_ERROR_INSUFFICIENT_MEMORY
```
Yes - will add these.

athoelke commented on doc/crypto/api/keys/policy.rst:
```diff
@@ -86,7 +86,18 @@ The usage flags are encoded in a bitmask, which has the type `psa_key_usage_t`.
+    -   `PSA_KEY_USAGE_VERIFY_HASH`
+    -   `PSA_KEY_USAGE_DERIVE`
+    -   `PSA_KEY_USAGE_VERIFY_DERIVATION`
+    -   `PSA_KEY_USAGE_ENCAPSULATE`
+    -   `PSA_KEY_USAGE_DECAPSULATE`
```
This seems reasonable.

Although ECIES uses an ECC key - the permitted algorithm policy would ensure that ENCRYPT usage indicates permission for `psa_encapsulate()`, and not `psa_asymmetric_encrypt()`. This is similar to the way ENCRYPT usage on an AES key is qualified by the permitted algorithm policy.

athoelke replied:
The [currently] last commit (https://github.com/ARM-software/psa-api/pull/222/commits/145c376df94631b360e914ca9d148156a4adc51e) has an implementation of this idea - removing the new usage flags and adding encapsulation and encapsulation to ENCRYPT/DECRYPT.

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+    Refer to the documentation of individual key-encapsulation algorithms for more information.
+
+    .. warning::
+        A :code:`PSA_SUCCESS` result from `psa_decapsulate()` does not guarantee that the output key is identical to the key produce by the call to `psa_encapsulate()`. For example:
```
> We should make it clear that _if the protocol is executed correctly_ then (at least with overwhelming probability) the two copies of the shared secret are identical.

I think this is what I had in mind, but the write-up lost some of the important clarifying text, leaving a very weak statement about what could be assumed.

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+            This is an input parameter: it is not updated with the final key attributes.
+            The final attributes of the new key can be queried by calling `psa_get_key_attributes()` with the key's identifier.
+    .. param:: psa_key_id_t * output_key
+        On success, an identifier for the newly created shared output key.
```
What about using "shared secret key" (this is the phrase used in FIPS203), as in the key with the shared secret?

I think 'shared output key' got into the text in order to disambiguate between the input key and the output key. In this API the input decapsulation key is `key`, so the output is `output_key`.In the key agreement API, the input key is called `private_key`, so the output is just `key`.

Perhaps use `decaps_key` as the input key, and then just `key` as the output - or maybe `shared_secret_key` to avoid confusion from truncated names?

gilles-peskine-arm replied:
I think `output_key` is ok as the parameter name. It's the English description that I find unclear.

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,350 @@
+        .. note::
+            Some key-encapsulation algorithms do not report an authentication failure explicitly.
+            Instead, an incorrect, pseudo-random key is output.
+    .. retval:: PSA_ERROR_INSUFFICIENT_MEMORY
```
Decapsulation does not require entropy (all provided previously in the decaps key generation, or by the other participant in the encapsulation).

However, storage errors are relevant for using the decaps key.

gilles-peskine-arm replied:
Decapsulation isn't randomized, but implementations may need entropy for blinding. We document `INSUFFICIENT_ENTROPY` for `psa_asymmetric_decrypt` which is in the same boat.

athoelke commented:
Addressed all of the issues raised - but one or two might benefit from validation and/or fine tuning. The removal of explicit encalsulation usage flags is a single commit that can be reverted if this does not turn out to be a better approach.

gilles-peskine-arm reviewed (changes requested):
LGTM except for one confusing paragraph.

athoelke commented on doc/crypto/api/ops/key-encapsulation.rst:
```diff
@@ -0,0 +1,381 @@
+Key-encapsulation algorithms are often referred to as 'key-encapsulation mechanisms' or KEMs.
+
+In a key-encapsulation algorithm, participant A generates a key-pair: a private decapsulation key, and a public encapsulation key.
+The public encapsulation key is made available to participant B, who needs to establish secure communication with the participant A.
```
Good point. I patterned the current text on material elsewhere, but the motivation context will also allow me to make that sentence simpler.

athoelke replied:
I've rewritten that paragraph into more of a flow:

> In a key-encapsulation algorithm, participants A and B establish a shared secret as follows:
>
>1.  Participant A generates a key-pair: a private decapsulation key, and a public encapsulation key.
>2.  The public encapsulation key is made available to participant B.
>3.  Participant B uses the encapsulation key to generate one copy of a shared secret, and some ciphertext.
>4.  The ciphertext is transferred to participant A.
>5.  Participant A uses the private decapsulation key to compute another copy of the shared secret.

athoelke commented:
Rebased following merge of #221, and added improvement for the description of key encapsulation.

gilles-peskine-arm reviewed (approved):
LGTM
